hyperon_das_atomdb_cpp/src/utils/expression_hasher.h

```diff
@@ -0,0 +1,173 @@
+
+#include <string>
+
+#include "type_aliases.h"
+
+using namespace std;
+
+namespace atomdb {
+
+#define JOINING_CHAR ((char) ' ')
+#define MAX_HASHABLE_STRING_SIZE ((size_t) 100000)
+#define MD5_BUFFER_SIZE ((size_t) 16)
+
+/**
+ * @class ExpressionHasher
+ * @brief A utility class for generating various types of hashes for expressions.
+ *
+ * The ExpressionHasher class provides static methods to compute MD5 hashes for different types of
+ * expressions, including named types, terminal expressions, composite expressions, and general
+ * expressions. It uses the mbedtls library for MD5 hashing.
+ *
+ * @note This class is designed to handle strings and lists of strings as input for hashing.
+ *
+ * @remark The class ensures that the generated hashes are consistent and unique for different
+ *         expressions, making it useful for scenarios where expression uniqueness and integrity
+ *         are critical.
+ */
+class ExpressionHasher {
+   public:
+    /**
+     * @brief Computes the MD5 hash of the given input string.
+     *
+     * @param input The input string to be hashed.
+     * @return A string representing the MD5 hash of the input.
+     */
+    static const string compute_hash(const string& input) {
+        auto ctx = unique_ptr<mbedtls_md5_context, decltype(&mbedtls_md5_free)>(
+            new mbedtls_md5_context(), &mbedtls_md5_free);
+        mbedtls_md5_init(ctx.get());
+        uchar md5_buffer[MD5_BUFFER_SIZE];
+        if (mbedtls_md5_starts_ret(ctx.get()) != 0 or
+            mbedtls_md5_update_ret(ctx.get(), (const uchar*) input.c_str(), input.length()) != 0 or
+            mbedtls_md5_finish_ret(ctx.get(), md5_buffer) != 0) {
+            throw runtime_error("Failed to compute MD5 hash");
+        }
+        char hash[2 * MD5_BUFFER_SIZE + 1];
+        for (unsigned int i = 0; i < MD5_BUFFER_SIZE; i++) {
+            sprintf(hash + 2 * i, "%02x", md5_buffer[i]);
+        }
+        hash[2 * MD5_BUFFER_SIZE] = '\0';
+        return move(string(hash));
+    }
+
+    /**
+     * @brief Generates a hash for a named type.
+     *
+     * @param name The name of the type.
+     * @return A string representing the hash of the named type.
+     */
+    static const string named_type_hash(const string& name) { return compute_hash(name); }
+
+    /**
+     * @brief Generates a hash for a terminal expression.
+     *
+     * @param type The type of the terminal expression.
+     * @param name The name of the terminal expression.
+     * @return A string representing the hash of the terminal expression.
+     * @throws invalid_argument if the terminal name is too large.
+     */
+    static const string terminal_hash(const string& type, const string& name) {
+        if (type.length() + name.length() >= MAX_HASHABLE_STRING_SIZE) {
+            throw invalid_argument("Invalid (too large) terminal name");
+        }
+        string hashable_string = type + JOINING_CHAR + name;
+        return compute_hash(hashable_string);
```

Comment:
I guess that putting a `move(...)` there will just reinforce what we expect from the compiler wiseness.
As this is a `return` statement from another function return that calls `move(...)`, AFAIK, the gcc compiler will implicitly make moves.
Anyways, adding explicit `move(...)` won't hurt and can prevent future problems in case of changes on the underlying functions. I will do it.

Reply:
Done.